src/lib/clean.c

```diff
@@ -57,61 +54,42 @@ static bool aa_hook_context_has_cache(AaHookContext *self, const char *d_name)
  */
 bool aa_hook_context_clean_cache(AaHookContext *self)
 {
-        autofree(DIR) *dir = NULL;
-        struct dirent *ent = NULL;
-        bool ret = false;
-        int fd = 0;
+        FTSENT* curr_entry = NULL;
+        char* path[] = { self->cache_dir, NULL };
+        FTS* file_system = fts_open(path, FTS_PHYSICAL | FTS_NOSTAT, NULL);
 
-        if (aa_unlikely(!self)) {
+        if (file_system == NULL) {
                 return false;
         }
 
-        /* If the directory doesn't exist yet, it isn't fatal */
-        dir = opendir(self->cache_dir);
-        if (!dir) {
-                if (errno == ENOENT) {
-                        return true;
+        bool ret = true;
+        while ((curr_entry = fts_read(file_system)) != NULL && errno == 0) {
+                switch (curr_entry->fts_info) {
+                case FTS_F:
+                case FTS_NSOK:
+                        /* Ignore the .features file and profiles that are still installed */
+                        if (strcmp(curr_entry->fts_name, ".features") == 0 ||
+                            aa_hook_context_has_cache(self, curr_entry->fts_name)) {
+                                continue;
+                        }
+                case FTS_DP:
+                        /* Remove the file or the dir if it's empty */
+                        if (remove(curr_entry->fts_path) == 0) {
+                                /* Print for benefit of calling tool */
+                                fprintf(stdout,
+                                        "aa_hook_context_clean_cache(): Removed %s\n",
+                                        curr_entry->fts_name);
+                        } else if (errno != ENOTEMPTY) {
```

Comment:
If this else if is checking the condition of the previous if(remove), it should not be an else if.